File: app/src/test/java/com/mohsenoid/rickandmorty/test/CharacterDataFactory.java
package com.mohsenoid.rickandmorty.test;

import com.mohsenoid.rickandmorty.data.db.dto.DbCharacterModel;
import com.mohsenoid.rickandmorty.data.network.dto.NetworkCharacterModel;
import com.mohsenoid.rickandmorty.domain.entity.CharacterEntity;

import java.util.ArrayList;
import java.util.List;

public class CharacterDataFactory {

    private CharacterDataFactory() { /* this will prevent making a new object of this type */ }

    public static class Db {

        public static DbCharacterModel makeDbCharacterModelWithIdAndStatus(int characterId, String status, Boolean isAlive, Boolean isKilledByUser) {
            return new DbCharacterModel(
                    characterId,
                    DataFactory.randomString(),
                    status,
                    isAlive,
                    DataFactory.randomString(),
                    DataFactory.randomString(),
                    DataFactory.randomString(),
                    OriginDataFactory.Db.makeDbOriginModel(),
                    LocationDataFactory.Db.makeDbLocationModel(),
                    DataFactory.randomString(),
                    DataFactory.randomString() + "," + DataFactory.randomString() + "," + DataFactory.randomString(),
                    DataFactory.randomString(),
                    DataFactory.randomString(),
                    isKilledByUser
            );
        }

        public static DbCharacterModel makeDbCharacterModelWithId(int characterId) {
            return makeDbCharacterModelWithIdAndStatus(characterId, DataFactory.randomString(), DataFactory.randomBoolean(), false);
        }

        public static DbCharacterModel makeDbCharacterModel() {
            return makeDbCharacterModelWithId(DataFactory.randomInt());
        }

        public static List<DbCharacterModel> makeDbCharactersModelList(int count) {
            List<DbCharacterModel> characters = new ArrayList<>();

            for (int i = 0; i < count; i++) {
                DbCharacterModel character = makeDbCharacterModel();
                characters.add(character);
            }

            return characters;
        }
    }

    public static class Network {

        public static NetworkCharacterModel makeNetworkCharacterModelWithId(int characterId) {
            return new NetworkCharacterModel(
                    characterId,
                    DataFactory.randomString(),
                    DataFactory.randomString(),
                    DataFactory.randomString(),
                    DataFactory.randomString(),
                    DataFactory.randomString(),
                    OriginDataFactory.Network.makeNetworkOriginModel(),
                    LocationDataFactory.Network.makeNetworkLocationModel(),
                    DataFactory.randomString(),
                    DataFactory.randomStringList(5),
                    DataFactory.randomString(),
                    DataFactory.randomString()
            );
        }

        public static NetworkCharacterModel makeNetworkCharacterModel() {
            return makeNetworkCharacterModelWithId(DataFactory.randomInt());
        }

        public static List<NetworkCharacterModel> makeNetworkCharactersModelList(int count) {
            List<NetworkCharacterModel> characters = new ArrayList<>();

            for (int i = 0; i < count; i++) {
                NetworkCharacterModel character = makeNetworkCharacterModel();
                characters.add(character);
            }

            return characters;
        }
    }

    public static class Entity {

        public static CharacterEntity makeCharacterEntityWithIdAndStatus(int characterId, String status, Boolean isAlive, Boolean isKilledByUser) {
            return new CharacterEntity(
                    characterId,
                    DataFactory.randomString(),
                    status,
                    isAlive,
                    DataFactory.randomString(),
                    DataFactory.randomString(),
                    DataFactory.randomString(),
                    OriginDataFactory.Entity.makeOriginEntity(),
                    LocationDataFactory.Entity.makeLocationEntity(),
                    DataFactory.randomString(),
                    DataFactory.randomIntList(5),
                    DataFactory.randomString(),
                    DataFactory.randomString(),
                    isKilledByUser
            );
        }

        public static CharacterEntity makeCharacterEntityWithId(int characterId) {
            return makeCharacterEntityWithIdAndStatus(characterId, DataFactory.randomString(), DataFactory.randomBoolean(), false);
        }

        public static CharacterEntity makeCharacterEntity() {
            return makeCharacterEntityWithId(DataFactory.randomInt());
        }

        public static List<CharacterEntity> makeEntityCharactersModelList(int count) {
            List<CharacterEntity> characters = new ArrayList<>();

            for (int i = 0; i < count; i++) {
                CharacterEntity character = makeCharacterEntity();
                characters.add(character);
            }

            return characters;
        }
    }
}

File: app/src/test/kotlin/com/mohsenoid/rickandmorty/test/CharacterDataFactory.kt
package com.mohsenoid.rickandmorty.test

import com.mohsenoid.rickandmorty.data.api.model.ApiCharacter
import com.mohsenoid.rickandmorty.data.db.model.DbCharacter
import com.mohsenoid.rickandmorty.domain.model.ModelCharacter

object CharacterDataFactory {

    internal fun makeDbCharacter(
        characterId: Int = DataFactory.randomInt(),
        status: String = DataFactory.randomString(),
        isAlive: Boolean = DataFactory.randomBoolean(),
        isKilledByUser: Boolean = false,
    ): DbCharacter =
        DbCharacter(
            id = characterId,
            name = DataFactory.randomString(),
            status = status,
            isAlive = isAlive,
            species = DataFactory.randomString(),
            type = DataFactory.randomString(),
            gender = DataFactory.randomString(),
            origin = OriginDataFactory.makeDbOrigin(),
            location = LocationDataFactory.makeDbLocation(),
            image = DataFactory.randomString(),
            episodeIds = DataFactory.randomIntList(count = 5),
            url = DataFactory.randomString(),
            created = DataFactory.randomString(),
            isKilledByUser = isKilledByUser,
        )

    internal fun makeDbCharacters(count: Int): List<DbCharacter> {
        val characters: MutableList<DbCharacter> = ArrayList()
        repeat(count) {
            val character: DbCharacter = makeDbCharacter()
            characters.add(character)
        }
        return characters
    }

    internal fun makeApiCharacter(characterId: Int = DataFactory.randomInt()): ApiCharacter =
        ApiCharacter(
            id = characterId,
            name = DataFactory.randomString(),
            status = DataFactory.randomString(),
            species = DataFactory.randomString(),
            type = DataFactory.randomString(),
            gender = DataFactory.randomString(),
            origin = OriginDataFactory.makeApiOrigin(),
            location = LocationDataFactory.makeApiLocation(),
            image = DataFactory.randomString(),
            episodes = DataFactory.randomIntList(count = 5)
                .map { "${DataFactory.randomString()}/$it" },
            url = DataFactory.randomString(),
            created = DataFactory.randomString(),
        )

    internal fun makeApiCharacters(count: Int): List<ApiCharacter> {
        val characters: MutableList<ApiCharacter> = ArrayList()
        repeat(count) {
            val character: ApiCharacter = makeApiCharacter()
            characters.add(character)
        }
        return characters
    }

    internal fun makeCharacter(
        id: Int = DataFactory.randomInt(),
        status: String = DataFactory.randomString(),
        isAlive: Boolean = DataFactory.randomBoolean(),
        isKilledByUser: Boolean = false,
    ): ModelCharacter = ModelCharacter(
        id = id,
        name = DataFactory.randomString(),
        status = status,
        isAlive = isAlive,
        species = DataFactory.randomString(),
        type = DataFactory.randomString(),
        gender = DataFactory.randomString(),
        origin = OriginDataFactory.makeOrigin(),
        location = LocationDataFactory.makeLocation(),
        imageUrl = DataFactory.randomString(),
        episodeIds = DataFactory.randomIntList(count = 5),
        url = DataFactory.randomString(),
        created = DataFactory.randomString(),
        isKilledByUser = isKilledByUser,
    )

    internal fun makeCharacters(count: Int): List<ModelCharacter> {
        val characters: MutableList<ModelCharacter> = ArrayList()
        repeat(count) {
            val character: ModelCharacter = makeCharacter()
            characters.add(character)
        }
        return characters
    }
}
